### DISCUSSION TITLE
FPrime for a distributed computing system?

### DISCUSSION THREAD

**Discussion by zimri-leisher** [nasa/fprime#2506] [Q&A] [answered]
Hello,
From my understanding, there is no support for running FPrime in a distributed computing system, where FPrime applications running on different computers would talk to each other. This makes sense because FPrime is tailored for small-scale spaceflight systems. But at the same time, the fact that components already have asynchronous queues for calling ports means that it might not be so bad to split one deployment into several applications that communicate via IP. It seems conceivable, although it would be a lot of work.
How crazy is this idea? Would it be better to just implement this at the user-level (as opposed to the framework level)?

**Comment by SterlingPeet** [comment, marked as answer]
This is not crazy at all, in fact Mars Helicopter uses multiple F Prime deployments to operate.  Usually we talk about using a [Hub Pattern](https://nasa.github.io/fprime/v1.5/UsersGuide/best/hub-pattern.html) to ship port calls between different deployments, which may run on different processors or even platforms.

If you wanted to try to abstract the entire port call mechanism to be able to cross processor and platform boundaries, that would be an interesting and complicated endeavor.  However, it isn't necessary to do that, if you want to try to run an intra-net of F Prime deployments.  It just is not common to do this with more than a small handful, so you might be blazing a new path if you wanted to use it as an HPC framework.

**Reply by LeStarch** [reply]
We made a `GenericHub` component: https://github.com/nasa/fprime/blob/devel/Svc/GenericHub/docs/sdd.md that supports this pattern.  With a reliable communication system for the underlying bytes, the hub can be extended pretty far!

**Comment by zimri-leisher** [comment]
Gotcha, that's super helpful! The Hub pattern might be the solution here. Abstracting the port call mechanism is tantalizing because it seems like it should be possible, but I guess it's better to go with what clearly works.

**Comment by ljreder** [comment]
Yes the so called Hub pattern was first conceptualized 10 years ago on some R&D effort going on at JPL.  The more interesting thing to me it is it possible to abstract away the so called hub components and create a model that looks like a single logical deployment where you via the modeling and build system move functionality to various processors and processor cores.  Here is a copy of the IEEE paper on it
[5620a054.pdf](https://github.com/nasa/fprime/files/14103925/5620a054.pdf).  This was back in the day when F′ was actually called ISF.  But you can wrap virtually any middleware in a lite weight F′ component and make some pretty nice open source reusable libraries doing this.
